src/lib/unlang/xlat_eval.c

```diff
@@ -222,6 +222,121 @@ static inline void xlat_debug_log_result(request_t *request, fr_value_box_t cons
+ * @param[in,out] in		value boxes provided as input.
+ * @param[in] input_type	required by xlat.
+ * @param[in] args		definition of arguments required by xlat.
+ */
+static xlat_action_t xlat_process_args(TALLOC_CTX *ctx, fr_value_box_list_t *in, request_t *request,
+					xlat_input_type_t input_type, xlat_arg_parser_t args[])
+{
+	xlat_arg_parser_t	*arg;
+	fr_value_box_t		*vb, *next;
+
+	/*
+	 *	No args registered for this xlat
+	 */
+	if (!args) return XLAT_ACTION_DONE;
+
+	switch (input_type) {
+	case XLAT_INPUT_UNPROCESSED:
+		/*
+		 *	xlat needs no input processing just return.
+		 */
+		return XLAT_ACTION_DONE;
+
+	case XLAT_INPUT_MONO:
+		/*
+		 *	xlat takes all input as a single vb.
+		 */
+		vb = fr_dlist_head(in);
+		if (vb) {
+			if (args->concat) {
+				/* Concat all vbs to one, casting to the required type. */
+				if (fr_value_box_list_concat(ctx, vb, in, args->type, true) < 0) {
+concat_error:
+					RERROR("Failed concatenating input");
+					return XLAT_ACTION_FAIL;
+				}
+			} else {
+				if ((vb->type == args->type) && (vb->type != FR_TYPE_VALUE_BOX)) return XLAT_ACTION_DONE;
+
+				/* Just cast the first vb to the required type */
+				if (fr_value_box_cast_in_place(ctx, vb, args->type, NULL) < 0) {
+cast_error:
+					RERROR("Failed converting input to %s",
+						fr_table_str_by_value(fr_value_box_type_table, args->type, "<INVALID>"));
+					return XLAT_ACTION_FAIL;
+				}
+			}
+		} else {
+			if (args->required) {
+missing:
+				RERROR("Missing required argument");
+				return XLAT_ACTION_FAIL;
+			}
+		}
+
+		break;
+
+	case XLAT_INPUT_ARGS:
+		/*
+		 *	xlat consumes a sequence of arguments.
+		 */
+		arg = args;
+		vb = fr_dlist_head(in);
+		while (arg->type != FR_TYPE_INVALID) {
+			/*
+			 *	pre-advance, in case the vb is replaced during processing
+			 */
+			next = fr_dlist_next(in, vb);
+
+			if ((arg->required) && (!vb)) goto missing;
+
+			if (!vb) break;
+
+			// *todo* make this escape tainted value boxes
+			if ((arg->concat) && (vb->type == FR_TYPE_GROUP)) {
+				fr_value_box_t *new_vb = fr_value_box_alloc(ctx, arg->type, NULL, false);
+				if (fr_value_box_list_concat(ctx, new_vb, &vb->vb_group, arg->type, true) < 0) {
+					talloc_free(new_vb);
+					goto concat_error;
+				}
+				fr_dlist_replace(in, vb, new_vb);
+				talloc_free(vb);
+			} else if ((vb->type == FR_TYPE_GROUP) && (arg->type != FR_TYPE_GROUP)) {
+				fr_value_box_t *child_vb = NULL;
+				/*
+				 *	Not concatenating children, move them up a level and cast if needed.
+				 */
+				while ((child_vb = fr_dlist_pop_tail(&vb->vb_group))) {
+					fr_dlist_insert_after(in, vb, child_vb);
+					if ((child_vb->type != arg->type) && (arg->type != FR_TYPE_VALUE_BOX)) {
+						if (fr_value_box_cast_in_place(ctx, child_vb, arg->type, NULL) < 0)
+							goto cast_error;
+					}
+				}
+
+				fr_dlist_remove(in, vb);
+				talloc_free(vb);
+			} else if ((vb->type != arg->type) && (arg->type != FR_TYPE_VALUE_BOX)) {
```

Comment:
Hmm, boxed values can't contain boxed values.  Don't worry about ```arg->type != FR_TYPE_VALUE_BOX``` the cast function will catch the error.